### DISCUSSION TITLE
Find references/usage of SFC with setup script

### DISCUSSION THREAD

**Discussion by weilbith** [vuejs/language-tools#3244] [Q&A] [answered]
Hey 👋 
I'm wondering how to do this for a long time. For any "normal" code in any language, I can ask the language server for references to the symbol under the cursor. That works with Volar too, ofc. But in case I have a file open that include a SFC using a setup script, how do I find the references or usages of this component in the project? I don't know which symbol I could go to, to call the language server on it. Is there maybe a special server command you could call for that? I couldn't find anything. 

Since I stumble over this issue for quite some time in my daily work, I'm no asking here for help. Maybe someone could point me into the right direction?

**Comment by weilbith** [comment]
I see that there is a `volar.vue.findAllFileReferences` command. But I fail to call it. I make a `workspace/executeCommand` request with this payload: `{ command: "volar.vue.findAllFileReferences", arguments: { textDocument: { uri: "<my_uri>" }}}`
But it does nothing.

**Reply by rchl** [reply, marked as answer]
It's not implemented through `workspace/executeCommand` but through a custom `volar/client/findFileReference` request that takes `{ textDocument: { uri: string } }` argument.

To be fair I'm not sure why custom requests are needed here rather than using the simpler to implement `workspace/executeCommand`

**Reply by weilbith** [reply]
Nice! That actually works. Thanks for that hint! 🙃

**Comment by crunchwrap89** [comment]
How would i be able to bind a shortcut for this? Specifically using webstorm

**Reply by weilbith** [reply]
That sounds very editor specific 😕 
I use NeoVim and there I could just implement it myself using Lua. Eventually I connected with the author of my favorite plugin to preview references and we added this as a feature.

Sorry that I can't help.

**Reply by Kahitar** [reply]
As a neovim user myself (a pretty new one), can you tell me how can I get this to work? 

Edit: For everyone interested, I used my trusted companion GPT4 to find the solution.

I put this at the top of my lsp-config.lua file (I use lazy and that file is in the plugins directory):
```lua
-- Custom function to send the volar/client/findFileReference request
function FindVueFileReferences()
    local bufnr = vim.api.nvim_get_current_buf()
    local uri = vim.uri_from_bufnr(bufnr)
    vim.lsp.buf_request(bufnr, 'volar/client/findFileReference', { textDocument = { uri = uri } }, function(err, result, ctx, config)
        if err then
            vim.notify("Error finding references: " .. vim.inspect(err), vim.log.levels.ERROR)
        else
            -- Convert the result to quickfix items
            local items = {}
            for _, ref in ipairs(result) do
                table.insert(items, {
                    filename = vim.uri_to_fname(ref.uri),
                    lnum = ref.range.start.line + 1,
                    col = ref.range.start.character + 1,
                    text = 'Reference'
                })
            end
            -- Populate the quickfix
            vim.fn.setqflist(items)
            -- Open the quickfix window
            vim.cmd('copen')
        end
    end)
end
```

And set the keymap:

```lua
vim.keymap.set('n', 'fr', '<cmd>lua FindVueFileReferences()<CR>', { noremap = true, silent = true })
```

**Reply by weilbith** [reply]
Well, there are plenty of ways to do it. That's NeoVim. 🤷🏾 
The basic request is the one like you used in your snippet. The probably most interesting and diverse part is how you'd like the result to be visualized. The quickfix list is a valid and nice solution. Very traditional.
I personally use the [glance.nvim](https://github.com/DNLHC/glance.nvim) plugin (see [this issue](https://github.com/DNLHC/glance.nvim/issues/56)) with a mapping only for Vue files. I like the plugin because of its look and feel. Fits nicely into my workflow. But there are plenty of other plugins. 🤷🏾

**Reply by mcchrish** [reply]
Here's how I did it with Neovim: https://github.com/mcchrish/dotfiles/blob/59a383188950ffca004d8d182789295615410d3f/vim/Library/Preferences/nvim/lua/mcchrish/plugins/lsp.lua#L117-L143

You can just basically use the builtin neovim LSP helpers, and it's pretty easy to pass things around.

But also now it's not working in v2. You have to stay with v1.18.17.

**Reply by Kahitar** [reply]
Thanks for the suggestions, very helpful. Will see if they improve something for me over my snippet I posted previously.

**Reply by nicobao** [reply]
Thank you for the ideas. I am trying to integrate this with Telescope. Has anyone already done it? If yes, could you share a snippet?

EDIT: that was in fact fairly easy. Simply replace `require("fzf-lua").quickfix()` with `require('telescope.builtin').quickfix()` and reuse the rest of the code here: 
https://github.com/mcchrish/dotfiles/blob/59a383188950ffca004d8d182789295615410d3f/vim/Library/Preferences/nvim/lua/mcchrish/plugins/lsp.lua#L133

**Comment by d9k** [comment]
in VSCode: `Ctrl+Shift+P`: `TypeScript: Find File References` command worked for me

**Reply by BoYuanSu** [reply]
Using "Find File References" still has some issues. For example, if an SFC is re-exported through an index file, using "Find File References" on the SFC won't directly show the places where the component is actually used.
